### PR TITLE
Split the CTC forced aligment API tutorial into two tutorials

### PR DESCRIPTION
**Pull request by xiaohui-zhang**
Splitting the multilingual example part into another tutorial.

### CODE REVIEW

File: docs/source/index.rst [hunk 1]
```diff
@@ -69,6 +69,7 @@ model implementations and application components.
    tutorials/online_asr_tutorial
    tutorials/device_asr
    tutorials/ctc_forced_alignment_api_tutorial
+   tutorials/forced_alignment_for_multilingual_data_tutorial
```
Comment (mthrok) [inline docs/source/index.rst:72]: Please move the original tutorial to `API Tutorials` section.

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+    print(
+        "Failed to import the forced alignment API. "
+        "Please install torchaudio nightly builds. "
+        "Please refer to https://pytorch.org/get-started/locally "
+        "for instructions to install a nightly build."
+    )
+    raise
+
+######################################################################
+# I. Preparation
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:43]: We don't use roman numerals. Please use arabic numerals or just simply remove them.

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+
+# %matplotlib inline
+from dataclasses import dataclass
+
+import matplotlib
+import matplotlib.pyplot as plt
+import torchaudio.functional as F
+from IPython.display import Audio
+
+matplotlib.rcParams["figure.figsize"] = [16.0, 4.8]
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:60]:
```suggestion
```

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+    def __repr__(self):
+        return f"{self.label}\t({self.score:4.2f}): [{self.start:5d}, {self.end:5d})"
+
+    @property
+    def length(self):
+        return self.end - self.start
+
+
+# compute frame alignments using torchaudio's forced alignment API
+def compute_alignments(transcript, dictionary, emission):
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:92]: Since the three helper functions, `compute_alignments`, `merge_repeats`, `merge_words` are already covered in the API tutorial, please merge these helper functions into one. We do not need to repeat calls to the same helper functions in this tutorial. This tutorial should forcus on "multi lingual" portion.

**Review by mthrok — COMMENTED** [review]
I'd like to see a lot more languages than just two.

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py (hunk 4 above)
Comment (xiaohui-zhang) [reply examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:92]: sure

**Comment by xiaohui-zhang** [comment]
> 

will add more

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py (hunk 2 above)
Comment (xiaohui-zhang) [reply examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:43]: sure

File: docs/source/index.rst (hunk 1 above)
Comment (xiaohui-zhang) [reply docs/source/index.rst:72]: sure

**Comment by xiaohui-zhang** [comment]
addressed comments

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+}
+
+
+######################################################################
+# Before aligning the speech with transcripts, we need to make sure
+# the transcripts are already romanized. The steps to romanize transcripts
+# using the "uroman" tool have been
+# listed below in comments.
+#
+
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:287]: This section can be organized better. What makes it hard to understand is that Python script used at the very end is at the beginning and it is followed up by the mixture of bash script and Python.

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+#
+# git clone https://github.com/isi-nlp/uroman
+# uroman/bin/uroman.pl < test.txt > test_romanized.txt
+#
+# file = "test_romanized.txt"
+# f = open(file, "r")
+# lines = f.readlines()
+# text_normalized = normalize_uroman(lines[0].strip())
+
+
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:320]: Use the code block for Python specific. Also move the definition of `normalize_uroman` here.

```
# .. code-block:: python
#
#    def normalize_uroman(text):
#         ...
#
#    file = ...
```

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+######################################################################
+# German example:
+# ~~~~~~~~~~~~~~~~
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:323]: For each language example, please print the original (un-romanized) sentence altogether. Then add the audio samples of segments. (make it more interactive)

Basically, to show what this API can do, include all the necessary information in one place.

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+waveform, _ = torchaudio.load(SPEECH_FILE)
+
+emission, waveform = get_emission(waveform)
+assert len(dictionary) == emission.shape[1]
+
+transcript = text_normalized
+
+segments, word_segments = compute_alignments(transcript, dictionary, emission)
+plot_alignments(segments, word_segments, waveform[0], emission.shape[0])
+plt.show()
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:339]: I believe plt.show is not required, and it will prohibit the tutorial code from being able to run as Python as regular script. Can you remove it?

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+# echo '关 服务 高端 产品 仍 处于 供不应求 的 局面' > test.txt
+#
+# Polish example:
+# echo 'wtedy ujrzałem na jego brzuchu okrągłą czarną ranę dlaczego mi nie powiedziałeś szepnąłem ze łzami' > test.txt
+#
+# Portuguese example:
+# echo 'mas na imensa extensão onde se esconde o inconsciente imortal só me responde um bramido um queixume e nada mais' > test.txt
+#
+# Italian example:
+# echo "elle giacean per terra tutte quante fuor d'una ch'a seder si levò ratto ch'ella ci vide passarsi davante" > test.txt
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:310]: Please move these examples to dedicated subsections.

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+    encoder_dropout=0.0,
+    encoder_layer_norm_first=True,
+    encoder_layer_drop=0.1,
+    aux_num_out=31,
+)
+
+torch.hub.download_url_to_file(
+    "https://dl.fbaipublicfiles.com/mms/torchaudio/ctc_alignment_mling_uroman/model.pt", "model.pt"
+)
+checkpoint = torch.load("model.pt", map_location="cpu")
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:221]: Can we use `load_state_dict_from_url`? so that the tutorial won't leave intermediate files here and there.

https://pytorch.org/docs/stable/hub.html#torch.hub.load_state_dict_from_url

File: examples/tutorials/ctc_forced_alignment_api_tutorial.py [hunk 11]
```diff
@@ -43,7 +43,7 @@
     raise
 
 ######################################################################
-# I. Basic usages
+# Basic usages
 # ---------------
```
Comment (mthrok) [inline examples/tutorials/ctc_forced_alignment_api_tutorial.py:47]:
```suggestion
# Basic usages
# ------------
```

File: examples/tutorials/ctc_forced_alignment_api_tutorial.py [hunk 12]
```diff
@@ -488,7 +488,7 @@ def display_segment(i, waveform, word_segments, frame_alignment):
 
 
 ######################################################################
-# II. Advanced usage: Dealing with missing transcripts using the <star> token
+# Advanced usage: Dealing with missing transcripts using the <star> token
 # ---------------------------------------------------------------------------
```
Comment (mthrok) [inline examples/tutorials/ctc_forced_alignment_api_tutorial.py]:
```suggestion
# Advanced usage: Dealing with missing transcripts using the <star> token
# -----------------------------------------------------------------------
```

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 13]
```diff
@@ -40,7 +40,7 @@
     raise
 
 ######################################################################
-# I. Preparation
+# Preparation
 # --------------
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py]:
```suggestion
# Preparation
# -----------
```

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 14]
```diff
@@ -193,11 +179,11 @@ def plot_alignments(segments, word_segments, waveform, input_lengths, scale=10):
 
 
 ######################################################################
-# II. Aligning non-English data
+# Aligning non-English data
 # -----------------------------
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py]:
```suggestion
# Aligning non-English data
# --------------------------
```

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 15]
```diff
@@ -193,11 +179,11 @@ def plot_alignments(segments, word_segments, waveform, input_lengths, scale=10):
 
 
 ######################################################################
-# II. Aligning non-English data
+# Aligning non-English data
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:199]: Do we need to emphasis *non*-English part?

File: examples/tutorials/ctc_forced_alignment_api_tutorial.py [hunk 16]
```diff
@@ -487,199 +488,60 @@ def display_segment(i, waveform, word_segments, frame_alignment):
...
-
-
-waveform, _ = torchaudio.load(SPEECH_FILE)
-
-
-def get_emission(waveform):
-    # NOTE: this step is essential
-    waveform = torch.nn.functional.layer_norm(waveform, waveform.shape)
+# Advanced usage: Dealing with missing transcripts using the <star> token
+# ---------------------------------------------------------------------------
```
Comment (mthrok) [inline examples/tutorials/ctc_forced_alignment_api_tutorial.py:492]:
```suggestion
# Advanced usage: Dealing with missing transcripts using the <star> token
# -----------------------------------------------------------------------
```

File: examples/tutorials/ctc_forced_alignment_api_tutorial.py [hunk 17]
```diff
@@ -487,199 +488,60 @@ def display_segment(i, waveform, word_segments, frame_alignment):
...
-    "'": 28,
-    "q": 29,
-    "x": 30,
-    "*": 31,
-}
+    emissions = torch.cat((emissions.cpu(), extra_dim), 2)
+    emission = emissions[0].detach()
+
+# Extend the dictionary to include the <star> token.
+dictionary["*"] = 29
```
Comment (mthrok) [inline examples/tutorials/ctc_forced_alignment_api_tutorial.py:517]: What happens to the regular use case if the start token is present in the dictionary?
If it does not have an effect, I think it's simpler to have these at the first time emission is computed and dictionary is constructed.

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 18]
```diff
@@ -0,0 +1,905 @@
...
+# Preparation
+# --------------
+#
+# Here we import necessary packages, and define utility functions for
+# computing the frame-level alignments (using the API
+# ``functional.forced_align``), token-level and word-level alignments, and
+# also alignment visualization utilities.
+#
+
+# %matplotlib inline
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:52]: Why do you need this?

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 19]
```diff
@@ -0,0 +1,905 @@
...
+# German example:
+# ~~~~~~~~~~~~~~~~
+
+text_raw = (
+    "aber seit ich bei ihnen das brot hole brauch ich viel weniger schulze wandte sich ab die kinder taten ihm leid"
+)
+text_normalized = (
+    "aber seit ich bei ihnen das brot hole brauch ich viel weniger schulze wandte sich ab die kinder taten ihm leid"
+)
+SPEECH_FILE = torchaudio.utils.download_asset("tutorial-assets/10349_8674_000087.flac")
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:346]:
```suggestion
SPEECH_FILE = torchaudio.utils.download_asset("tutorial-assets/10349_8674_000087.flac", progress=False)
```

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 20]
```diff
@@ -0,0 +1,905 @@
...
+# tokenization (separated by spaces) in its raw written form. In order to
+# obtain word level alignments, you need to first tokenize the transcripts
+# at the word level using a word tokenizer like `“Stanford
+# Tokenizer” <https://michelleful.github.io/code-blog/2015/09/10/parsing-chinese-with-stanford/>`__.
+# However this is not needed if you only want character-level alignments.
+#
+
+text_raw = "关 服务 高端 产品 仍 处于 供不应求 的 局面"
+text_normalized = "guan fuwu gaoduan chanpin reng chuyu gongbuyingqiu de jumian"
+SPEECH_FILE = torchaudio.utils.download_asset("tutorial-assets/mvdr/clean_speech.wav")
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:483]:
```suggestion
SPEECH_FILE = torchaudio.utils.download_asset("tutorial-assets/mvdr/clean_speech.wav", progress=False)
```

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py (hunk 19 above)
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:346]: `SPEECH_FILE` in other tutorials are all capped like global variables to indicate that it is used throughout the tutorial, not just in one section.

Here, and from this point, SPEECH_FILE is overwritten multiple times, which looks strange. Can you lower case?

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py (hunk 20 above)
Comment (mthrok) [reply examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:483]: Same about the SPEECH_FILE variable name.

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 21]
```diff
@@ -0,0 +1,905 @@
...
+
+
+######################################################################
+# Polish example:
+# ~~~~~~~~~~~~~~~
+
+
+text_raw = "wtedy ujrzałem na jego brzuchu okrągłą czarną ranę dlaczego mi nie powiedziałeś szepnąłem ze łzami"
+text_normalized = "wtedy ujrzalem na jego brzuchu okragla czarna rane dlaczego mi nie powiedziales szepnalem ze lzami"
+SPEECH_FILE = torchaudio.utils.download_asset("tutorial-assets/5090_1447_000088.flac")
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:552]:
```suggestion
SPEECH_FILE = torchaudio.utils.download_asset("tutorial-assets/5090_1447_000088.flac", progress=False)
```

Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:552]: same about the `SPEECH_FILE` variable name

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 22]
```diff
@@ -0,0 +1,905 @@
...
+# ~~~~~~~~~~~~~~~~~~~
+
+
+text_raw = (
+    "mas na imensa extensão onde se esconde o inconsciente imortal só me responde um bramido um queixume e nada mais"
+)
+text_normalized = (
+    "mas na imensa extensao onde se esconde o inconsciente imortal so me responde um bramido um queixume e nada mais"
+)
+SPEECH_FILE = torchaudio.utils.download_asset("tutorial-assets/6566_5323_000027.flac")
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:655]: ditto
```suggestion
SPEECH_FILE = torchaudio.utils.download_asset("tutorial-assets/6566_5323_000027.flac", progress=False)
```

File: examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py [hunk 23]
```diff
@@ -0,0 +1,905 @@
...
+
+######################################################################
+# Italian example:
+# ~~~~~~~~~~~~~~~~
+
+text_raw = "elle giacean per terra tutte quante fuor d'una ch'a seder si levò ratto ch'ella ci vide passarsi davante"
+text_normalized = (
+    "elle giacean per terra tutte quante fuor d'una ch'a seder si levo ratto ch'ella ci vide passarsi davante"
+)
+SPEECH_FILE = torchaudio.utils.download_asset("tutorial-assets/642_529_000025.flac")
```
Comment (mthrok) [inline examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:780]:
```suggestion
SPEECH_FILE = torchaudio.utils.download_asset("tutorial-assets/642_529_000025.flac", progress=False)
```

Comment (mthrok) [reply examples/tutorials/forced_alignment_for_multilingual_data_tutorial.py:780]: same about the variable name

**Review by mthrok — APPROVED** [review]
